### PR TITLE
OSASINFRA-3620: Add manila csi

### PR DESCRIPTION
**Pull request by MaysaMacedo**
This is the second part of the process to merge OpenStack Manila CSI operator into csi-operator.

This PR does the following:

- Adds Manila assets
- Adds the assets generator implementation
- Copy Manila Secret controller
- Use the control plane and guest namespaces
- Adds a new controller that manages the start of all controllers
- Adapt manila to use the precondition available with the new controller

### CODE REVIEW

File: pkg/operator/config/config.go [hunk 1]
```diff
@@ -71,6 +71,11 @@ type OperatorControllerConfig struct {
 
 	// ExtraReplacements defines additional replacements that should be made to assets
 	ExtraReplacementsFunc func() []string
+
+	// Precondition to run operator
+	Precondition func() (bool, error)
+
+	PreconditionInformers []factory.Informer
```
Comment (stephenfin) [inline pkg/operator/config/config.go:78]:
```suggestion
	// List of informers that should be added to the precondition controller.
	PreconditionInformers []factory.Informer
```

File: pkg/operator/starter_controller.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+const (
+	// Minimal interval between controller resyncs.
+	resyncInterval = 20 * time.Minute
+
+	operatorConditionPrefix = "CSIController"
+)
+
+func StarterController(operatorClient v1helpers.OperatorClient,
```
Comment (stephenfin) [inline pkg/operator/starter_controller.go:38]: Could we get a doc string explaining what this is doing? fwiw, it is checking that the preconditions, if any, are met and then starting/stopping every other controller accordingly?

File: pkg/operator/starter_controller.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	if c.operatorControllerConfig.Precondition != nil {
+		_, err := c.operatorControllerConfig.Precondition()
+		if err != nil {
+			return c.setDisabledCondition(ctx, fmt.Sprintf("Pre condition not satisfied: %v", err))
```
Comment (stephenfin) [inline pkg/operator/starter_controller.go:69]: nit:

```suggestion

			return c.setDisabledCondition(ctx, fmt.Sprintf("Pre-condition not satisfied: %v", err))
```

File: pkg/operator/starter_controller.go
Comment (stephenfin) [inline pkg/operator/starter_controller.go:1]: IMO this should probably be the first commit in the PR. Heck, it's probably large enough to be a separate PR. Putting it first highlights that this is a "groundwork" change though.

File: pkg/driver/openstack-manila/openstack.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	}
+
+	provider, err := openstack.NewClient(opts.IdentityEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a provider client: %w", err)
+	}
+
+	// we represent version using commits since we don't tag releases
+	ua := gophercloud.UserAgent{}
+	ua.Prepend(fmt.Sprintf("csi-driver-manila-operator/%s", version.Get().GitCommit))
```
Comment (stephenfin) [inline pkg/driver/openstack-manila/openstack.go:58]:
```suggestion
	ua.Prepend(fmt.Sprintf("openstack-manila-csi-driver-operator/%s", version.Get().GitCommit))
```

File: pkg/driver/openstack-manila/openstack.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	cloud, err := getCloudFromFile(cloudConfigFilename)
+	if err != nil {
+		return nil, err
+	}
+	return &openStackClient{
+		cloud: cloud,
+	}, nil
+}
+
+func (o *openStackClient) GetShareTypes() ([]sharetypes.ShareType, error) {
```
Comment (stephenfin) [inline pkg/driver/openstack-manila/openstack.go:36]: I would like to rewrite all of this to just use `clouds.yaml` but that's a job for later...

File: pkg/driver/openstack-manila/openstack.go
Comment (stephenfin) [inline pkg/driver/openstack-manila/openstack.go:1]: nit: This file feels like it belongs in `pkg/openstack-manila` since it's not exclusively related to the driver functionality.

File: pkg/driver/openstack-manila/openstack_manila.go [hunk 6]
```diff
@@ -265,6 +265,12 @@ func generateStorageClass(shareType sharetypes.ShareType, guestNamespace string)
 	return sc
 }
 
+// withClusterWideProxyHook adds the cluster-wide proxy config to the DaemonSet.
+func withClusterWideProxyDaemonSetHook(_ *clients.Clients) (csidrivernodeservicecontroller.DaemonSetHookFunc, []factory.Informer) {
+	hook := csidrivernodeservicecontroller.WithObservedProxyDaemonSetHook()
+	return hook, nil
+}
+
```
Comment (stephenfin) [inline pkg/driver/openstack-manila/openstack_manila.go:273]: This could probably do with being rolled back into the original commit?

**Review by stephenfin — CHANGES_REQUESTED** [review]
Lots of comments. Some of them nits, some of them a bit larger. Note that I went through this commit-by-commit so some might only make sense in that context. Please do ask if anything doesn't make sense though

File: pkg/driver/common/operator/replacer.go [hunk 7]
```diff
@@ -64,6 +65,12 @@ func DefaultReplacements(controlPlaneNamespace, guestNamespace string) []string
 	if csiDriver != "" {
 		pairs = append(pairs, []string{"${HYPERSHIFT_IMAGE}", hyperShiftImage}...)
 	}
+
+	nfsImage := os.Getenv(nfsImageEnvName)
+	if nfsImage != "" {
+		pairs = append(pairs, []string{"${NFS_DRIVER_IMAGE}", nfsImage}...)
+	}
+
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/driver/common/operator/replacer.go:73]: indeed, thanks

File: pkg/driver/openstack-manila/openstack_manila.go [hunk 8]
```diff
@@ -94,6 +94,7 @@ func GetOpenStackManilaGeneratorConfig() *generator.CSIDriverGeneratorConfig {
 func GetOpenStackManilaOperatorConfig() *config.OperatorConfig {
 	return &config.OperatorConfig{
 		CSIDriverName:                   opv1.ManilaCSIDriver,
+		GuestNamespace:                  "openshift-manila-csi-driver",
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/driver/openstack-manila/openstack_manila.go:108]: sure thing

File: pkg/operator/starter_controller.go (hunk 2 above)
Comment (MaysaMacedo) [reply pkg/operator/starter_controller.go:38]: Yes, that's the point of this controller

File: pkg/driver/openstack-manila/openstack_manila.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	}
+	cfg := operator.NewDefaultOperatorControllerConfig(flavour, c, "OpenStackManila")
+
+	go c.ConfigInformers.Start(ctx.Done())
+
+	// Hooks to run on all clusters
+	cfg.AddDeploymentHookBuilders(c, withCABundleDeploymentHook)
+	cfg.AddDaemonSetHookBuilders(c, withCABundleDaemonSetHook)
+
+	cfg.DeploymentWatchedSecretNames = append(cfg.DeploymentWatchedSecretNames, metricsCertSecretName)
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/driver/openstack-manila/openstack_manila.go]: I don't think we need to ensure the Pod restarts. The secret is not mounted in any Manila Pod.
We have instead reference to the secrets in the [volumesnapshotclass](assets/overlays/openstack-manila/base/volumesnapshotclass.yaml) and in the generation of the [StorageClass](https://github.com/openshift/csi-operator/pull/288/files#diff-60e4d8814f67651d7a754bb0d38e0c6afdf700fd60c475a806fca524ff92176aR276-R281).
I updated the clouds.yaml with the following command:
`oc set data -n kube-system secret/openstack-credentials clouds.yaml="$(<"$ALTERNATIVE_CLOUDS_YAML")"`
and I noticed the `secrets/csi-manila-secrets -n openshift-manila-csi-driver` was also updated immediately.

**Comment by MaysaMacedo** [comment]
@stephenfin This PR is ready for another look whenever you can. Thanks

File: pkg/openstack-manila/secret/secretsync.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+)
+
+// This SecretSyncController translates Secret provided by cloud-credential-operator into
+// format required by the CSI driver.
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/openstack-manila/secret/secretsync.go:25]: I will give a try to remove it in a separate PR

File: pkg/driver/openstack-manila/openstack_manila.go [hunk 11]
```diff
@@ -149,6 +150,7 @@ func GetOpenStackManilaOperatorControllerConfig(ctx context.Context, flavour gen
 
 // withCABundleDeploymentHook projects custom CA bundle ConfigMap into the CSI driver container
 func withCABundleDeploymentHook(c *clients.Clients) (dc.DeploymentHookFunc, []factory.Informer) {
+	klog.Info("withCABundleDeploymentHook")
```
[the remark this answers is not in the document]
Comment (stephenfin) [reply pkg/driver/openstack-manila/openstack_manila.go]: nit: this is still here. If you have to rebase you should probably delete it here in commit `openstack-manila: configure guest namespace`.

**Review by stephenfin — COMMENTED** [review]
Still one small nit in there but that's not enough to respin this for and the rest of my comments have been addressed. Nice work.

/lgtm

File: pkg/operator/starter.go [hunk 12]
```diff
@@ -214,15 +221,16 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
...
-	for _, controller := range csiOperatorControllerConfig.ExtraControlPlaneControllers {
-		klog.Infof("Starting controller %s", controller.Name())
-		go controller.Run(ctx, 1)
-	}
-	klog.Info("Starting control plane controllerset")
-	go controlPlaneCSIControllerSet.Run(ctx, 1)
-	klog.Info("Starting guest controllerset")
-	go guestCSIControllerSet.Run(ctx, 1)
+	starterController := StarterController(
+		c.OperatorClient,
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/operator/starter.go:225]: This is not necessarily delaying the startup. If there is no precondition defined, which is the case of all drivers besides manila, this controller will immediately start all the other controllers, if they haven't started yet.
Every 20 mins this precondition will be checked, but as there is no precondition for the other drivers and the controllers are already running at that point, there is nothing to do. 

This design was discussed back in the day with @jsafrane, it was the only way to make this change generic enough to not have a different controller for Manila only.

File: pkg/clients/builder.go [hunk 13]
```diff
@@ -42,16 +42,14 @@ type Builder struct {
...
-		csiDriverName:    csiDriverName,
-		guestNamespace:   guestNamespace,
-		resync:           resync,
-		controllerConfig: controllerConfig,
-		controlPlaneNamespaces: []string{
-			controllerConfig.OperatorNamespace,
-		},
+		userAgwent:             userAgent,
+		csiDriverName:          csiDriverName,
+		guestNamespace:         guestNamespace,
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/clients/builder.go:49]: The `guestNamespaces` is used to add informers to those namespace in [here](https://github.com/openshift/csi-operator/blob/master/pkg/clients/builder.go#L121). Maybe that could be renamed to `guestNamespacesToInform`? or maybe just adding a comment is enough?

File: pkg/operator/starter.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (gnufied) [reply pkg/operator/starter.go:225]: I am personally worried about regressions in this code path for operators that don't need it. If `PreCondition` is nil - then those controllers will not be started until `Sync` function within `starterController` turns.

File: pkg/driver/openstack-manila/openstack_manila.go [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+		if err != nil {
+			return false, err
+		}
+
+		if len(shareTypes) == 0 {
+			klog.V(4).Infof("Manila does not provide any share types")
+			return false, errors.New("manila does not provide any share types")
+		}
+
+		syncCSIDriver(ctx, c.KubeClient, c.EventRecorder)
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply pkg/driver/openstack-manila/openstack_manila.go:175]: AFAIK storage classes needs to be created with the share types that exists, that is why `syncStorageClasses` needs to be done regularly in a controller.

**Comment by rlobillo** [comment]
Hello @MaysaMacedo 

openstack-tests includes a test to confirm that manila is working fine. However, it is skipped:

```
skip [github.com/openshift/openstack-test/test/extended/openstack/volumes.go:192]: No StorageClass with manila.csi.openstack.org provisioner
Ginkgo exit error 3: exit with code 3
```

Do you know if there is any job in the CI that uses a manila-enabled openstack cloud so the test is run?

**Comment by rlobillo** [comment]
/qe-approved

According to https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_csi-operator/288/pull-ci-openshift-csi-operator-master-e2e-openstack-manila-csi/1853875164861698048/artifacts/e2e-openstack-manila-csi/openshift-e2e-test/artifacts/e2e.log results.
